dali/operators/reader/loader/tar_utils.cc

```diff
@@ -0,0 +1,189 @@
+  ParseHeader();
+}
+
+TarArchive::TarArchive(TarArchive&& other) {
+  *this = static_cast<TarArchive&&>(other);
+}
+
+TarArchive::~TarArchive() {
+  Unregister(instance_handle);
+}
+
+TarArchive& TarArchive::operator=(TarArchive&& other) {
+  stream = std::move(other.stream);
+  eof = other.eof;
+  filename = other.filename;
+  filesize = other.filesize;
+  readoffset = other.readoffset;
+  archiveoffset = other.archiveoffset;
+
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  instances[instance_handle] = nullptr;
+  instance_handle = other.instance_handle;
+  return *this;
+}
+
+uint64_t RoundToBlockSize(uint64_t count) {
+  return ((count + kBlockSize - 1) / kBlockSize) * kBlockSize;
+}
+
+bool TarArchive::NextFile() {
+  if (eof) {
+    return false;
+  }
+  Skip(RoundToBlockSize(filesize) - readoffset);
+  eof = ParseHeader();
+  return !eof;
+}
+
+bool TarArchive::IsAtFile() const {
+  return !eof;
+}
+
+std::string TarArchive::GetFileName() const {
+  return filename;
+}
+
+uint64_t TarArchive::GetFileSize() const {
+  return filesize;
+}
+
+std::shared_ptr<uint8_t> TarArchive::ReadFile() {
+  archiveoffset -= readoffset;
+  readoffset = 0;
+  stream->Seek(archiveoffset);
+  auto buffer = std::shared_ptr<uint8_t>(new uint8_t[filesize], [](uint8_t* ptr) { delete[] ptr; });
+  Read(buffer.get(), filesize);
+  return buffer;
+}
+
+uint64_t TarArchive::Read(uint8_t* buffer, uint64_t count) {
+  if (eof) {
+    return 0;
+  }
+  count = std::max(std::min(count, filesize - readoffset), 0_u64);
+  uint64_t num_read_bytes = stream->Read(buffer, count);
+  readoffset += num_read_bytes;
+  archiveoffset += num_read_bytes;
+  return count;
+}
+
+bool TarArchive::Eof() const {
+  return readoffset >= filesize;
+}
+
+inline void TarArchive::Skip(int64 count) {
+  stream->Seek(archiveoffset += count);
+  readoffset += count;
+}
+
+int LibtarOpenTarArchive(const char*, int oflags, ...) {
+  va_list args;
+  va_start(args, oflags);
+  const int instance_handle = va_arg(args, int);
+  va_end(args);
+  return instance_handle;
+}
+
+
+inline bool TarArchive::ParseHeader() {
+  TAR* handle;
+  tartype_t type = {LibtarOpenTarArchive, [](int) -> int { return 0; },
+                    [](int instance_handle, void* buf, size_t count) -> ssize_t {
+                      const auto current_archive = instances[instance_handle];
+                      const ssize_t num_read =
+                          current_archive->stream->Read(reinterpret_cast<uint8_t*>(buf), count);
+                      current_archive->archiveoffset += num_read;
+                      return num_read;
+                    },
+                    [](int, const void*, size_t) -> ssize_t { return 0; }};
+  tar_open(&handle, "", &type, 0, instance_handle, TAR_GNU);
+
+  int errorcode = th_read(handle);
+  if (errorcode) {
+    filename = "";
+    filesize = 0;
+    if (errorcode == -1) {
+      DALI_FAIL(R"(Corrupted tar file at )" + handle->pathname);
```

Comment:
```suggestion
      DALI_FAIL("Corrupted tar file at " + handle->pathname);
```
No need for a raw string here.

Reply:
alright